polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+    if (this.constantEra) calendarDate = { ...calendarDate, era: this.constantEra };
+    return { ...calendarDate, year, eraYear, month, monthCode: `M${month}` };
+  },
+  regulateMonthDayNaive(calendarDate, overflow, cache) {
+    const largestMonth = this.monthsInYear(calendarDate, cache);
+    let { month, day } = calendarDate;
+    if (overflow === 'reject') {
+      if (month < 1 || month > largestMonth) throw new RangeError(`Invalid month: ${month}`);
+      if (day < 1 || day > this.maximumMonthLength(calendarDate)) throw new RangeError(`Invalid day: ${day}`);
+    } else {
+      if (month < 1) month = 1;
+      if (month > largestMonth) month = largestMonth;
+      if (day < 1) day = 1;
+      if (day > this.maximumMonthLength(calendarDate)) day = this.maximumMonthLength(calendarDate);
+    }
+    return { ...calendarDate, month, day };
+  },
+  calendarToIsoDate(date, overflow = 'constrain', cache) {
+    const originalDate = date;
+    // First, normalize the calendar date to ensure that (year, month, day)
+    // are all present, converting monthCode and eraYear if needed.
+    date = this.adjustCalendarDate(date, cache, overflow, false);
+
+    // Fix obviously out-of-bounds values. Values that are valid generally, but
+    // not in this particular year, will be handled lower below.
+    date = this.regulateMonthDayNaive(date, overflow, cache);
+
+    const { year, month, day } = date;
+    const key = JSON.stringify({ func: 'calendarToIsoDate', year, month, day, overflow, id: this.id });
+    let cached = cache.get(key);
+    if (cached) return cached;
+    // If YMD are present in the input but the input has been constrained
+    // already, then cache both the original value and the constrained value.
+    let keyOriginal;
+    if (
+      originalDate.year !== undefined &&
+      originalDate.month !== undefined &&
+      originalDate.day !== undefined &&
+      (originalDate.year !== date.year || originalDate.month !== date.month || originalDate.day !== date.day)
+    ) {
+      keyOriginal = JSON.stringify({
+        func: 'calendarToIsoDate',
+        year: originalDate.year,
+        month: originalDate.month,
+        day: originalDate.day,
+        overflow,
+        id: this.id
+      });
+      cached = cache.get(keyOriginal);
+      if (cached) return cached;
+    }
+
+    // First, try to roughly guess the result
+    let isoEstimate = this.estimateIsoDate({ year, month, day });
+    const calculateSameMonthResult = (diffDays) => {
+      // If the estimate is in the same year & month as the target, then we can
+      // calculate the result exactly and short-circuit any additional logic.
+      // This optimization assumes that months are continuous (no skipped days)
+      // so it breaks in the case of calendar changes like 1582's skipping of
+      // days in October in the Roman calendar. This is just a prototype
+      // implementation so this error is OK for now.
```

Comment:
As per my experience, the ICU `japanese` calendar uses the Julian calendar representation of dates before 1582-10-15.
Oct. 4 10 Tensho corresponds to 1582-10-14, and one day later is Oct. 15 Tensho. The same hold for `buddhist` and `roc`. You can make a quick online test using [this moke-up](https://louis-aime.github.io/calendrical-javascript/dateextendtest-en). So effectively the proposed optimisation will break for those calendar.

Reply:
I consider those skipped days in `japanese`, `buddhist`, and `roc` to be a bug because they are different from `gregory` which uses the proleptic calendar.  For this reason I filed https://bugs.chromium.org/p/chromium/issues/detail?id=1173158 which will hopefully align `japanese`, `buddhist`, and `roc` to match `gregory` behavior.

That said, the optimization above isn't actually broken for any month in those calendars except the month that contains the skipped days. If the month has no skipped days, then the optimization works fine.

There's a *different* problem with those calendars which is that their results don't match Gregorian for pre-ISO1582 dates.  This will also be resolved by the V8 bugfix above.

Reply:
> None of the ICU calendars are affected by this / have skipped days, right?

This will be correct, assuming that the V8 bug linked above is resolved as I think it should be. ;-)